### ISSUE TITLE
Issues when pulling images in parallel with aufs

### ISSUE THREAD

**Issue by fermayo** [docker/docker#15283]
Description of problem: When pulling images in parallel from the same base layer (one from a v1 registry, and another one from a v2 registry), docker panics and the layer is never downloaded/extracted properly

`docker version`:
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d
OS/Arch (server): linux/amd64

`docker info`:
Containers: 28
Images: 318
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 399
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-44-generic
Operating System: Ubuntu 14.04.1 LTS
CPUs: 2
Total Memory: 7.797 GiB
Name: ip-10-77-100-61
ID: VHB6:RP2I:XGVB:EQRG:KPAH:JB7E:VWYA:53Z4:2WPL:7LNB:UKUP:PXDM
WARNING: No swap limit support

`uname -a`:
Linux ip-10-77-100-61 3.13.0-44-generic #73-Ubuntu SMP Tue Dec 16 00:22:43 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.): AWS

How reproducible: 3 out of 3 tries

Steps to reproduce: pulling two images that share layers in parallel, one in a v2 registry and another one in a v1 registry.

Actual Results: https://gist.github.com/fermayo/449f166c1d784e3a9300

Expected Results: no panic

Additional info:

**Comment by e-max** [comment]
We frequently face  problems connected with parallel pull and/or different versions of registry as well. Although we use btrfs.   

Usually it looks like one of two situations:
* docker pull starts download layers, but when it meet some layer it suddenly stops and print "image doesn't exits"

* docker pull finishes correctly but all files  added in some layer just disappeared.

**Comment by LK4D4** [comment]
ping @dmcgowan @stevvooe

**Comment by stevvooe** [comment]
This looks like a duplicate of #9718. It is possible that the issue is not fully fixed. #14193 provided mitigation but we have not been able to identify the root cause. Perhaps closing #9718 was pre mature.

The panic should be fixed by the @jlhawn patch.

Can you reproduce this with [1.8.0-rc3](https://github.com/docker/docker/releases/tag/v1.8.0-rc3)? I suspect the panic will be fixed but the layer corruption may still be present.

**Comment by e-max** [comment]
I've reproduced "Image not found" behavior on 1.8.0-rc3.  
```
root@kube-minion-2:~# docker pull registry-db.test.ivi.ru/geoip:production_20150807
production_20150807: Pulling from geoip
428b411c28f0: Pulling fs layer 
435050075b3f: Layer already being pulled by another client. Waiting. 
9fd3c8c9af32: Pulling fs layer 
6d4946999d4f: Pulling fs layer 
0cb330f72706: Pulling fs layer 
dacb2f9367e1: Pulling fs layer 
6e9327e94b73: Pulling fs layer 
4c8b6bd004bd: Pulling fs layer 
6a1e5b5181b8: Pulling fs layer 
fec5d0e2df6c: Pulling fs layer 
ca0f04b2cdda: Pulling fs layer 
7c0a4794c5e2: Pulling fs layer 
2733bbd9a2ab: Pulling fs layer 
71e5db383a1e: Pulling fs layer 
5f3a2b96b106: Pulling fs layer 
4ed1000f08ea: Pulling fs layer 
083e9e8989c0: Pulling fs layer 
46a333b54eeb: Pulling fs layer 
26a37a293eda: Pulling fs layer 
Pulling repository registry-db.test.ivi.ru/geoip
Error: image geoip:production_20150807 not found
```

```
DEBU[6123] Calling GET /images/{name:.*}/json           
INFO[6123] GET /images/registry-db.test.ivi.ru/geoip:production_20150807/json 
ERRO[6123] Handler for GET /images/{name:.*}/json returned error: No such image: registry-db.test.ivi.ru/geoip:production_20150807 
ERRO[6123] HTTP Error                                    err=No such image: registry-db.test.ivi.ru/geoip:production_20150807 statusCode=404
```

**Comment by e-max** [comment]
Full log

```
DEBU[6655] Calling POST /images/create                  
INFO[6655] POST /v1.20/images/create?fromImage=registry-db.test.ivi.ru%2Fgeoip%3Aproduction_20150807 
DEBU[6655] hostDir: /etc/docker/certs.d/registry-db.test.ivi.ru 
DEBU[6655] Trying to pull registry-db.test.ivi.ru/geoip from https://registry-db.test.ivi.ru v2 
DEBU[6655] Pulling tag from V2 registry: "production_20150807" 
DEBU[6655] Verification failed for /geoip using key STO3:KY26:7IY7:VPBZ:BHIY:JEHK:2UL4:ZTEL:WWRX:FXCW:7GPC:CMKK 
DEBU[6655] Key check result: not verified               
DEBU[6655] Verification failed for /geoip using key STO3:KY26:7IY7:VPBZ:BHIY:JEHK:2UL4:ZTEL:WWRX:FXCW:7GPC:CMKK 
DEBU[6655] Key check result: not verified               
DEBU[6655] pulling blob "sha256:6ffe5d2d6a9775ef70b5a636aac1175f73f556ee82c6457600959bdeded94cb1" to 428b411c28f0c33e561a95400a729552db578aee0553f87053b96fc0008cca6a 
DEBU[6655] pulling blob "sha256:f4e00f994fd4f90d9eafda2f7b3f9ea984599c50d14e608f0a5ec9bb9e365b10" to 435050075b3f881611b0f4c141bb723f38603caacd31a13a185c1a38acfb4ade 
DEBU[6655] pulling blob "sha256:e99f3d1fc87b485492190612bf949108bd8b25f5299d3ffd16846c6e52760fd7" to 9fd3c8c9af32dddb1793ccb5f6535e12d735eacae16f8f8c4214f42f33fe3d29 
DEBU[6655] pulling blob "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" to 6d4946999d4fb403f40e151ecbd13cb866da125431eb1df0cdfd4dc72674e3c6 
DEBU[6655] pulling blob "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" to 0cb330f72706e8b3865d5d161171171920b60ab40849aafc30caca3c940eae86 
DEBU[6655] pulling blob "sha256:31e4406b561355280963a3bbf22a6f2b91165ae0689ad2e979d5971a1deb5135" to dacb2f9367e1d0e88941f7b1fa5ac303632670a2b42d618d575128294f59dfff 
DEBU[6655] pulling blob "sha256:ca6cef90130b4d95fd435b8c077a60d77c5f0e1d4f5e8038162b8d9573766b8e" to 6e9327e94b73b080a20291a99bd1804b21f73e21bf507cb40729ee9277bd4f62 
DEBU[6655] pulling blob "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" to 4c8b6bd004bd8997eec518886d7e42237937e3013b21bf52ac19d6e92e65961e 
DEBU[6655] pulling blob "sha256:952ebe903641922d40669cd67e23281159c69a0deb77362259b37f0e52a0798c" to 6a1e5b5181b8a1562a4787916fd86bfbdce879f96a4c46c6121e13748a4944fb 
DEBU[6655] pulling blob "sha256:02750888125ab4a4f76f5bd2df729ab6cbb2d9f5ade852b9a6853ec268f78934" to fec5d0e2df6c36d093db604f2607b0672ee7a4d4c9444a124d96eb03fbb544e2 
DEBU[6655] pulling blob "sha256:a5526fd57a1fc41d893fddcfd5ed90a58ad97070554c4d618298abc360a27dc4" to ca0f04b2cdda9058442c9fdc8facb0fba6f73018f9ceeab98a7cd012475f8a6f 
DEBU[6655] pulling blob "sha256:0e1095947af0d45ebcb7f670d71b28af3cfc9c3ac12c058b1ea53141767a0062" to 7c0a4794c5e20c771f7da0cf6f729dc29575329d07eb0eaf295f3384cee20300 
DEBU[6655] pulling blob "sha256:91bc4d66679e351f64464cc8b1633c38940d56640824a301f875201c26597e96" to 2733bbd9a2ab9140bbdf7625db737e38e3547215d4242053cfb199935439ea0e 
DEBU[6655] pulling blob "sha256:0930e95622db5b480f25d090e914c9eb2cdf0a4da5f8c4e0f935a29d5c26ad86" to 71e5db383a1ed2f868e023d70c20c7dd0a376d3f46c1b568db10beac7f9ad928 
DEBU[6655] pulling blob "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" to 5f3a2b96b1066a5a0a917f8fdc2c8e435c34eef8e9656769f524dd74ea451686 
DEBU[6655] pulling blob "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" to 4ed1000f08ea19d8988f897695637a8f0dcced13c0edbfd7f9c70bb3ceaebb05 
DEBU[6655] pulling blob "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" to 083e9e8989c0efad9e9b5895751ecaede86f534a4d00889d55368bf5e0598da7 
DEBU[6655] pulling blob "sha256:ba3befe982d8df237325005020f2abcb17b04c3279b46ea4ca694df8c33cc6b7" to 46a333b54eeb0335c1c31916c906c2efff1a9e68ce0982a2387ca02371727c12 
DEBU[6655] pulling blob "sha256:5f84137dfa70501730dbd5d53db004b5e2791d593e894efc81c28e62407314c4" to 26a37a293edae13df4f3d0ecb41ed3ddeb902da62c04d82b4d786a787a7b6b43 
DEBU[6655] pulling blob "sha256:5f84137dfa70501730dbd5d53db004b5e2791d593e894efc81c28e62407314c4" to 26a37a293edae13df4f3d0ecb41ed3ddeb902da62c04d82b4d786a787a7b6b43 
DEBU[6655] Error trying v2 registry: open /var/lib/docker/tmp/GetImageBlob769439480: no such file or directory 
DEBU[6655] Trying to pull registry-db.test.ivi.ru/geoip from https://registry-db.test.ivi.ru v1 
DEBU[6655] hostDir: /etc/docker/certs.d/registry-db.test.ivi.ru 
DEBU[6655] attempting v2 ping for registry endpoint https://registry-db.test.ivi.ru/v2/ 
ERRO[6655] error copying from layer download progress reader: download canceled 
DEBU[6655] Endpoint https://registry-db.test.ivi.ru/v2/ is eligible for private registry. Enabling decorator. 
DEBU[6655] [registry] Calling GET https://registry-db.test.ivi.ru/v1/repositories/geoip/images 
DEBU[6655] Not continuing with error: Error: image geoip:production_20150807 not found 
```

**Comment by stevvooe** [comment]
@e-max Thank you for the report. Had you seen the panic before? We think the panic is actually fixed.

I've reopened #9718, since this problem wasn't actually fixed.

**Comment by e-max** [comment]
No.  No panic.   Thanks.  Is there any additional information or test I can provide in order to help with this issue ?

**Comment by stevvooe** [comment]
@e-max At least we fixed the panic. ;)

This might be a lot to ask, but it might help if we could get an strace of the docker daemon during the download process. My theory is that the image GC is being overzealous and issuing a remove while the download is still ongoing. If we can catch the extra remove (in your case, `/var/lib/docker/tmp/GetImageBlob769439480`), we may be able to identify where in the process the incorrect call is happening.

That said, the error messages you have are slightly different that that encountered by @fermayo and in  #9718, so we should a little open minded that this is a separate issue.

Thanks for sticking with us on this. Your patience and willingness to provide information really helps out in making this project better.

**Comment by pditommaso** [comment]
Same problem here. Using Docker 1.7.1 the pull process hangs. While using version 1.8.1 the client returns the following error message: 

    Unable to find image 'xx/yy' locally
    Error response from daemon: No such image: xx/yy 

This is pretty a serious issue because makes impossible to having concurrent docker jobs to pull the same image.

Downgrading to 1.6.2 it works fine. Below you can find Docker & OS info 

	$uname -a
	Linux xxx 3.19.0-15-generic #15-Ubuntu SMP Thu Apr 16 23:32:37 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

	$ docker version
	Client:
	 Version:      1.8.1
	 API version:  1.20
	 Go version:   go1.4.2
	 Git commit:   d12ea79
	 Built:        Thu Aug 13 02:40:42 UTC 2015
	 OS/Arch:      linux/amd64

	Server:
	 Version:      1.8.1
	 API version:  1.20
	 Go version:   go1.4.2
	 Git commit:   d12ea79
	 Built:        Thu Aug 13 02:40:42 UTC 2015
	 OS/Arch:      linux/amd64
 
 
	$ docker -D info
	Containers: 0
	Images: 64
	Storage Driver: aufs
	 Root Dir: /var/lib/docker/aufs
	 Backing Filesystem: extfs
	 Dirs: 64
	 Dirperm1 Supported: true
	Execution Driver: native-0.2
	Logging Driver: json-file
	Kernel Version: 3.19.0-15-generic
	Operating System: Ubuntu 15.04
	CPUs: 12
	Total Memory: 188.9 GiB
	Name: russia.crg.es
	ID: EYGW:KNDC:Q6YQ:BBYQ:HUMV:2DUX:R3RI:MFNI:QIPD:5OZY:U5OQ:57FG
	WARNING: No swap limit support

**Comment by ncdc** [comment]
Is this still an issue? It sounds like it should be fixed in 1.9+?

**Comment by stevvooe** [comment]
@ncdc Agreed.

If anyone is still seeing this in 1.9, let us know and we'll investigate.
